src/SecretsFile.hs

```diff
@@ -0,0 +1,166 @@
+  show sfErr =
+    case sfErr of
+      IOError -> "could not read file"
+      ParseError pe -> MP.parseErrorPretty pe
+
+-- | Helper for ExceptT stuff that we use in app/Main.hs
+readSecretList :: (MonadError SFError m, MonadIO m) => FilePath -> m [Secret]
+readSecretList fp = liftEither =<< (liftIO $ readSecretsFile fp)
+
+-- | Read a lit of secrets from a file
+readSecretsFile :: FilePath -> IO (Either SFError [Secret])
+readSecretsFile fp = do
+  contents <- safeReadFile fp
+  case contents of
+    Just c -> do
+      let parseResult = parseSecretsFile fp c
+      case parseResult of
+        Right res -> pure $ Right res
+        Left err -> pure $ Left (ParseError err)
+    Nothing -> pure $ Left IOError
+
+-- | Read a file, catching all IOError exceptions.
+safeReadFile :: FilePath -> IO (Maybe String)
+safeReadFile fp = do
+  (contentsOrErr :: Either IOError String) <- (try . readFile) fp
+  case contentsOrErr of
+    Right contents -> pure $ Just contents
+    Left _ -> pure $ Nothing
+
+-- | Parse a String as a SecretsFile.
+parseSecretsFile :: FilePath -> String -> Either (MP.ParseError (MP.Token String) Void) [Secret]
+parseSecretsFile = MP.parse secretsFileP
+
+-- | SpaceConsumer parser, which is responsible for stripping all whitespace.
+--
+-- Sometimes, we require explicit newlines, therefore, we don't handle those
+-- here.
+whitespace :: Parser ()
+whitespace = MPL.space whitespaceChars lineComment blockComment
+  where
+    whitespaceChars = void $ MP.takeWhile1P (Just "whitespace") (\c -> isSpace c && c /= '\n')
+    lineComment = MP.empty
+    blockComment = MP.empty
+
+-- | Parses one or multiple newlines separated by whitespace.
+newlines :: Parser ()
+newlines = void $ some $ lexeme $ MPC.char '\n'
+
+-- | Helper which consumes all whitespace after a parser
+lexeme :: Parser a -> Parser a
+lexeme = MPL.lexeme whitespace
+
+-- | Helper which looks for a string and consumes trailing whitespace.
+symbol :: String -> Parser String
+symbol = MPL.symbol whitespace
+
+-- | Top level parser of the secrets file
+--
+-- Parses the magic version number and dispatches to the Mount block based
+-- parser or the list based parser based on that.
+secretsFileP :: Parser [Secret]
+secretsFileP = do
+  _ <- whitespace
+  version <- versionP
+  case version of
+    V1 -> some (secretP version "secret")
+    V2 -> concat <$> some secretBlockP
+
+-- | Parse the file version
+--
+-- We need @MP.try@ because we need to backtrack after reading VERSION. (As
+-- some secrets could very well start with that path.
+versionP :: Parser SFVersion
+versionP = option V1 $ MP.try $ do
+  _ <- symbol "VERSION"
+  symbol "2"
+  _ <- newlines
+  pure V2
+
+-- | Parse a secret block
+--
+-- Exclusive to V2 of the format. A secret block consists of a line describing
+-- the mount location followed by secret specifications.
+secretBlockP :: Parser [Secret]
+secretBlockP = do
+  _ <- symbol "MOUNT"
+  mountPath <- lexeme (some MPC.alphaNumChar)
```

Comment:
Is `alphaNumChar` what Vault supports? Maybe link to their docs here.

Reply:
This is more restrictive. They seem to support almost anything, except control characters.

I've lifted some restrictions, but we still impose the following:

 - No whitespace (would require some quoting and escaping mechanism)
 - No control characters
 - No `#` and `=` in path components. Allowing this would make the following things ambiguous:

```
FOO=foo#bar
foo#bar#baz
```